_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,239 @@
+This time is spent on loading of libraries, preparing, optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, [ccache](https://ccache.dev) 
+and [sccache](https://github.com/mozilla/sccache) are used.
+Let's look at the assembly scheme using C++ as an example
+to imagine the build process in compiled languages:
+
+<p align="center">
+  <img src="/images/defaultCPhase.svg">
+</p>
+
+1) First, preprocessor gets the input files. The input files are source files (.cpp) and header files (.h).
+The result is a single edited file with human-readable code that the compiler will get.
+
+
+2) The compiler receives the finished code file and converts it into machine code, presented in an object file.
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language.
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project are compiled in parallel.
+
+3)  Then, the linker gets object files.
+Linker is a program that combines object files into an executable file or library.
+The result of the linker is an executable .exe file.
+
+
+As a result, in compiled languages, multiple files are simultaneously and independently converted
+into machine code at the compilation stage.
+This machine code is then combined into one executable file.
+
+
+`ccache` uses hashcode to find cached files. The hashcode includes information:
+file contents, directory, compiler information, compilation time, extensions
+used by the compiler. A compressed machine code file is placed in the cache using the received key.
+
+
+`ccache` has two main caching methods:
+1) `Direct mode` - hashcode is generated based on the source code. 
+`Direct mode` compiles the program faster, since the preprocessor step is skipped.
+However,the header files are not checked for changes, so the wrong project may be built.
+2) `Preprocessor mode` - hashcode is generated based on the result of preprocessor.
+`Preprocessor mode` is slower than `direct mode`, but the right project is built always.
+
+`Sccache`, unlike `ccache`, allows you to store cached files not only locally, but also in the cloud.
+And it also has fixed some bugs (for example, there is a check of header files, which makes direct mode more accurate).
+
+
+### Maven
+[Maven](https://maven.apache.org) automates and manages Java-project builds. 
+Building a project in `Maven` is completed in three
+maven [LifeCycles Maven](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html),
+which consist of `phases`. `Phases` consist of sets of `goals`.
+
+`Maven` has default `phases` and `goals`  for building any projects:
+
+<p align="center">
+  <img src="/images/defaultPhaseMaven.svg">
+</p>
+
+In `Maven` all phases and goals are executed strictly in order, linearly.
```

Comment:
@Yanich96 So, Maven doesn't use caching at all?

Reply:
@volodya-lombrozo As far as I understand, that Maven can use added extensions from Gradle for caching. Or Maven can rebuild only changed project modules.

Reply:
@Yanich96 Maven has `.m2` folder at least. In this folder it keeps all downloaded dependencies. So it's some sort of caching too.

Reply:
@Yanich96 It's not about Gradle, I guess: https://maven.apache.org/extensions/maven-build-cache-extension/